feat: initial test template docs and major reorganisation (#336)
pactflow/docs.pactflow.io: website/docs/docs/ai/using.md

      parameters:
      - name: id
        in: path
        description: Product ID
        schema:
          type: string
        required: true
      responses:
        "200":
          description: successful operation
          content:
            "application/json; charset=utf-8":
              schema:
                $ref: '#/components/schemas/Product'
        "404":
          description: Product not found
          content: {}
components:
  schemas:
    Product:
      type: object
      required:
        - id
        - name
        - price
      properties:
        id:
          type: string
        type:
          type: string
        name:
          type: string
        version:
          type: string
        price:
          type: number
```

#### Generating Tests from OpenAPI Descriptions

OpenAPI Descriptions (OAD) are machine and human-readable descriptions of an API. As it includes details such as  resources, verbs, status codes and request/response bodies, it is a great source for Pact tests. Learn more about OpenAPI [here](https://swagger.io/docs/.specification/v3_0/about/).

Generate a Pact test for the HTTP 200 use case (default):

> ```sh
> pactflow-ai generate openapi \
>   --spec     /tmp/products.yml \
>   --endpoint "/product/{id}" \
>   --output   /tmp/api.pact.spec.ts \
>   --language typescript
> ```

Reviewer: ````suggestion
```console
pactflow-ai generate \
  --openapi     /tmp/products.yml \
  --endpoint "/product/{id}" \
  --output   /tmp/api.pact.spec.ts \
  --language typescript
```
````

**Testing Negative Scenarios**

By default, generation will focus on the "happy" paths and ignore the "error" paths. 

We can generate a test for the `404` scenario by explicitly specifying the methods and statuses we are interested in:

> ```sh
>  pactflow-ai generate openapi \
>   --spec ./products.yml \
>   --endpoint '{"path": "/product/{id}", "methods": ["GET"], "status": ["404"]}' \
>   --output ./src/api.pact.spec.ts \
>   --language typescript
> ```

Reviewer: ````suggestion
```console
 pactflow-ai generate \
  --openapi ./products.yml \
  --endpoint '{"path": "/product/{id}", "methods": ["GET"], "status": ["404"]}' \
  --output ./src/api.pact.spec.ts \
  --language typescript
```
````

Using these expressions, you have fine-grained control over the generation process.

> :::tip
> Refer to `pactflow-ai generate openapi --help` for more on the supported matching expressions.
> :::

Reviewer: ```suggestion
:::tip
Refer to `pactflow-ai generate --help` for more on the supported matching expressions.
:::
```

**Considerations**

As this form of generation does not have access to the client source code, tests generated will use a placeholder API client in place of the real API client, when calling the mock service in the test.

Be sure to update your tests to insert the real API client to ensure the tests are valid.

#### Generating Tests from Code

> If you already have an existing code base, using the `code` option can help to quickly bootstrap coverage of your contract tests.

Reviewer: ```suggestion
If you already have an existing code base, it is possible for the PactFlow AI to refer to it and help quickly bootstrap coverage of your contract tests.
```

**Example**

Let's assume you already have a client implemented and integrated with the Product API, and you have the following (simplified) API client code:

**api.js**
```js
export class API {
  // ...
  async getProduct(id) {
    return axios
      .get(this.withPath("/product/" + id), {
        headers: {
          Authorization: this.generateAuthToken(),
        },
      })
      .then((r) => new Product(r.data));
  }
}
```

> ...and the following `Product` definition:

Reviewer: ```suggestion
and the following `Product` definition:
```

**product.js**
```js
export class Product {
  constructor({id, name, price}) {
    this.id = id
    this.name = name
    this.price = price
  }
}
```

We can generate pact tests using the `pactflow-ai generate code` subcommand:

> ```sh
> pactflow-ai generate code \
>   --output ./src/api.pact.spec.ts \
>   --language typescript \
>   ./src/api.js \
>   ./src/product.js
> ```

Reviewer: ````suggestion
```console
pactflow-ai generate \
  --output ./src/api.pact.spec.ts \
  --language typescript \
  --code ./src/api.js \
  --code ./src/product.js
```
````

Note how we can pass in multiple files to the command. Providing the right information to PactFlow AI will ensure it has the context to generate usable output.

#### Generating Tests from your OpenAPI Descriptions + Code

> If you have both an OAD and code, you can pass both to the `openai` subcommand as follows:

Reviewer: ```suggestion
If you have both an OAD and code, they can be combined to generate a more informed Pact test targeting your codebase as follows:
```

> ```sh
> pactflow-ai generate openapi \
>   --spec ./products.yml \
>   --endpoint "/product/{id}" \
>   --output ./src/api.pact.spec.ts \
>   --language typescript \
>   --code ./src/product.js \
>   --code ./src/api.js
>  ```

Reviewer: ````suggestion
```console
pactflow-ai generate \
  --openapi ./products.yml \
  --endpoint "/product/{id}" \
  --output ./src/api.pact.spec.ts \
  --language typescript \
  --code ./src/product.js \
  --code ./src/api.js
```
````

This tends to provide the best output, as it has the most information about how the API works and how your code integrates with it.

#### Generating Tests from Request-Response

> In some cases, you may not have access to an OAD or code. You may already have a large test suite where you have - or can - record the request and response details. You may simply have detailed descriptions such as a BDD suite, describing how the solution should work.

Reviewer: ```suggestion
In some cases, you may not have access to an OAD or code. You may already have a large test suite where you have—or can—record the request and response details. You may simply have detailed descriptions such as a BDD suite, describing how the solution should work.
```

In these cases, you can create a suite of Pact tests from these recordings.

Given the following request-response pairs (taken from the output of `curl` command, and in the format of an [HTTP Message](https://developer.mozilla.org/en-US/docs/Web/HTTP/Messages)):

**get.request.http**
```
GET /product/10 HTTP/1.1
Host: api.example.com
User-Agent: curl/8.1.2
Authorization: Bearer notarealtoken
Accept: application/json; charset=UTF-8
```

**get.response.http**
```
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Content-Length: 96

{
  "id": "10",
  "name": "Aussie",
  "type": "Pizza",
  "version": "v1",
  "price": 9.99,
}
```

Run the following command to generate a Pact test:

> ```sh
>  pactflow-ai generate request-response \   # use the request-response subcommand
>   --request ./capture/get.request.http \   # path to the request description
>   --response ./capture/get.response.http \ # path to the response description
>   --language typescript \
>   --output ./src/api.pact.spec.ts
> ```

Reviewer: ````suggestion
```console
 pactflow-ai generate \   # use the request-response subcommand
  --request ./capture/get.request.http \   # path to the request description
  --response ./capture/get.response.http \ # path to the response description
  --language typescript \
  --output ./src/api.pact.spec.ts
```
````

**Considerations**

As this form of generation does not have access to the client source code, tests generated will use a placeholder API client in place of the real API client, when calling the mock service in the test.

> Be sure to update your tests to insert the real API client to ensure the tests are valid.

Reviewer: Once the unified endpoint/refactor we're doing is done, we'll be able to combine request/response with all other inputs.

One consideration about the implementation will be that the request-response is deemed to be _most specific_ and will drive what Pact test is going to be generated. For example, if combining a request-response with an OpenAPI doc, the AI will automatically determine what part of the OAD is relevant and trim the rest (in detail, it builds its own `--endpoint` matcher)

### Using Test Templates

Test Templates allow teams to generate contract tests that align with their existing style, frameworks, and SDK versions. By defining templates as code, users can ensure that generated tests match their project conventions from the start, reducing manual refactoring and improving efficiency.

Test Templates are supported for all forms of test generation.

:::warning
TBC - need to finish this section :)
:::

## Best Practices

### 1. Check PactFlow's work

While PactFlow AI is a powerful tool, it is still capable of making mistakes, and you should always validate the code it suggests. Use the following tips to ensure you are accepting accurate, secure suggestions:

1. Understand suggested code before you use it.
2. Ensure the tests are appropriately scoped and target your system under test (SUT), usually an API client package.

>    1. Make sure the tests _do not_ use a generic HTTP client. This is particularly relevant for `openapi` and `request-response` generated code.

Reviewer: ```suggestion
   1. Make sure the tests _do not_ use a generic HTTP client. This is particularly relevant for `openapi` and `request-response` generated code. Without knowing your codebase, a placeholder is used which needs to be replaced with the relevant calls to your codebase.
```

```

**get.response.http**
```
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Content-Length: 96

{
  "id": "10",
  "name": "Aussie",
  "type": "Pizza",
  "version": "v1",
  "price": 9.99,
}
```

Run the following command to generate a Pact test:

```console
pactflow-ai generate \
  --request  capture/get.request.http \   # path to the request description
  --response capture/get.response.http \ # path to the response description
  --language typescript \
  --output   src/api.pact.spec.ts
```

**Considerations**

As this form of generation does not have access to the client source code, tests generated will use a placeholder API client in place of the real API client, when calling the mock service in the test.

Be sure to update your tests to insert the real API client to ensure the tests are valid. 

As an example:

```js
return provider.executeTest(async (mockserver) => {
  const client = new ProductClient(mockserver.url);
  const response = await client.getProduct(10);
  expect(response.data).to.deep.equal({
    id: 27,
    storeId: '009111fc-992a-4cae-96d4-6507b657b0e4',
    price: 99.09,
    categories: ['hardware'],
  });
});
```

> You need to replace the 

Reviewer 2: There's something missing.

Author: Fixed, thanks!

#### Generating Tests from multiple sources

> If you have more than one source availabl, such as an OAD and code, you can pass  to the `generate` command as follows:

Reviewer 2: available*

When generating the output, the precedence from least-to-most specific impact on the output is as follows:

1. Request-response
2. Code
3. OpenAPI Document

For example, if combining a request-response with an OpenAPI doc, PactFlow AI will automatically determine what part of the OAD is relevant and trim the rest (effectively building its own `--endpoint` matcher).

### Using Test Templates

Test Templates allow teams to generate contract tests that align with their existing style, frameworks, and SDK versions. By defining templates as code or providing additional contextual prompts, users can ensure that generated tests match their project conventions from the start, reducing manual refactoring and improving efficiency.

Test Templates are supported for all forms of test generation.

#### Providing Code Templates

When generating code, you can ask PactFlow to use a template test to customise output. Using the template, PactFlow AI will attempt to use the style, conventions and patterns it sees in your code. To provide a template, use the `--test-template` parameter, passing the location of a file containing the template.

**Example:**

> The following is an example Node JS 

Reviewer 2: Just 'Node JS' will do.
